Doc/regions.md

```diff
@@ -0,0 +1,567 @@
+```
+>>> x = 1
+>>> y = 2
+>>> calc("4 x y + *")
+12
+```
+
+The numpy library is pretty popular, so this is a good example of how to use the frame object in a way that is not debugging.
+
+
+Note that the frame object is mostly read-only, but does contain an interesting operation to set the executing line number of a frame.
+```C
+static int
+frame_setlineno(PyFrameObject *f, PyObject* p_new_lineno, void *Py_UNUSED(ignored))
+```
+This seems very bad to call from another thread/interpreter.
+
+Ultimately, a frame object should never escape the thread/interpreter that it was created in.
+
+
+
+## Regions in Veronapy.
+
+The previous section illustrates the many temporary objects that Python creates to handle variables.
+To support this, we need our region object model to support complex local structures for indexing into regions.
+To achieve this, we will use three kinds of object ownership:
+
+* Local object 
+* Immutable
+* Regions
+
+The "local objects" are used to represent the many local structures that Python creates such as function objects, frame objects and cells.
+These objects are typically not shared between threads, so normally should not require any region tracking.
+However, we need to support local objects becoming immutable to allow them to be shared, and suppport for groups of objects to be moved between interpreters.
+
+To pass mutable states between threads, we need to introduce a new construct to Python: "regions".
+A region is a group of objects that can move together across threads
+and can be accessed by at most one interpreter at a time.
+This removes the need for locks to protect the objects in the region.
+The region can be passed between threads, and the receiving interpreter can access the objects in the region.
+
+To efficiently support regions, we impose the following invariants that the runtime is responsible for maintaining:
+
+* Local objects can reach anything
+* Immutable objects can only reach other immutable objects.
+* Objects in a region can only reach other objects in the same region, immutable objects, or linear region entry points.
+* Regions track the number of incoming references from local objects.
+
+New objects are always created as a local object.
+If Veronapy features are not used, then all the objects in the system will be local objects and the runtime will behave as normal.
+
+
+### Regions
+
+Regions are used to pass mutable state between concurrent units of execution (behaviours).
+For usability, we need references into the region, that are not part of the region itself.
+These might be from the objects such as the frame, from closures or from function objects.
+This is precisely what local objects are for as they can refer into the region.
+
+The core challenge is allowing local objects to refer into a region,
+while allowing the region to be sent to another interpreter.
+To do this, we need to track the number of references from local objects to a region, the "local reference count" (LRC).
```

Comment:
To be clear, local references to the region entry point are fine, right, and the LRC only covers local references directly *into* the region (bypassing the entry point), right? And the LRC is a single number per region?

Reply:
87dc94b8fda8648f6046b6d2cafdd575bc4bfaf8

Entry point still to handle